### PR TITLE
feat: aggregate values to features (DHIS2-9530)

### PR DESCRIPTION
**Pull request by turban**
Partly fixes: https://jira.dhis2.org/browse/DHIS2-9530

This PR makes a tighter integration between maps-gl and the JavaScript API of Google Earth Engine. It allows data to be aggregated to org unit features in various ways. The Earth Engine layer is also refactored to make the code cleaner and easier to test. 

if you want to test this branch: 
- Checkout `feat/DHIS2-9530-ee-refactor` branch of maps-app
- In maps-gl: `yarn install` / `yarn build` / `cd build` / `yarn link`
- In maps-app: `yarn link @dhis2/maps-gl`

There are two related PRs in maps app:
- https://github.com/dhis2/maps-app/pull/1476
- https://github.com/dhis2/maps-app/pull/1477

Main changes: 
- Org unit features can be included in the `data` option, but the layer will still show if we don't add them. 
- The org unit boundaries will be drawn above the EE image layer. The boundary will be highlighted on mouse hover.
- The EE image will be clipped to the org unit features if included.
- New method "aggregate" allowing the EE data to be aggregated to the org unit features. The calculations are happening on EE servers using the JavaScript API. We support aggregation of a single variable (e.g. population and elevation) and multiple/classed data (e.g. landcover). We also allow "bands" to be aggregated (e.g. sum different age groups before calculating the totals for the org.unit). 
- For classed data, we can calculate the percentage of the total org unit area, or calculate the area in hectares or acres.  
-  Layer option `"aggregation": "mosaic"` is changed to `"mosaic": true`. Main reason was to avoid confusion with all the aggregation logic. 
- Added new `"imageready"` event for EE layer - it means that the EE image is ready and can be used for calculations/aggregations. 
- Added on load event and callback for layers. So far it's only in used for EE layers, which often take time to load/process. Used to show a spinner while the layer is loading in the Maps app. 
- Switched from twoDecimals to numberPrecision utility function which can take variable number of decimals. 
- I've added unit tests for the EE layer and new utility functions. The test coverage is not complete.   
- Method `createdLegend` was removed, as most of the code was duplicated in the Maps app. 

After this PR: 

Screenshot 2021-02-16 at 18 37 21

Screenshot 2021-02-16 at 18 36 44

Screenshot 2021-02-16 at 18 36 10

Before this PR: 

Screenshot 2021-02-16 at 18 39 54

### CODE REVIEW

**Review by edoardo — APPROVED** [review]
I'm not able to comment on the implementation as I don't know enough about the whole engine.
For the code itself, I added a couple of minor comments, but overall looks fine.

File: src/layers/EarthEngine.js [hunk 1]
```diff
@@ -15,208 +25,241 @@ class EarthEngine extends Layer {
...
-                item.to = to
-                item.name = item.from + ' - ' + item.to
-            } else {
-                // Higher than max
-                item.from = from
-                item.name = '> ' + item.from
-            }
+    // Apply array of filters returns image
+    applyFilter = async datasetId => {
+        const { filter, mosaic } = this.options
```
[the remark this answers is not in the document]
Comment (turban) [reply src/layers/EarthEngine.js:180]: `filters` where already taken for ao, so I decided to use `filter` - could have taken a different name like `collectionFilters` or something. `filter` has been used for several versions so let's keep it for now :-)

File: src/layers/EarthEngine.js [hunk 2]
```diff
@@ -15,208 +25,241 @@ class EarthEngine extends Layer {
...
-            to = Math.round(min + step * (index + (min > 0 ? 1 : 2)))
+        let collection = ImageCollection(datasetId)
 
-            return item
+        // Scale is lost when creating a mosaic below
+        // https://developers.google.com/earth-engine/guides/projections
+        this.scale = await getScale(collection.first())
+
+        filter.forEach(f => {
+            collection = collection.filter(
```
[the remark this answers is not in the document]
Comment (turban) [reply src/layers/EarthEngine.js:196]: This is performed on the Earth Engine servers and not in the browsers - and it's the recommended way to extract images from a larger collection, like one period.

File: src/layers/EarthEngine.js [hunk 3]
```diff
@@ -15,208 +25,241 @@ class EarthEngine extends Layer {
...
+            // Image has classes (e.g. landcover)
+            this.params = getParamsFromLegend(legend)
+            return eeImage
+        }
+
+        const min = 0
+        const max = legend.length - 1
+        const { palette } = params
+
+        for (let i = min, item; i < max; i++) {
```
[the remark this answers is not in the document]
Comment (turban) [reply src/layers/EarthEngine.js:262]: Keeping it as max is legend.length - 1

File: src/layers/EarthEngine.js [hunk 4]
```diff
@@ -225,70 +268,53 @@ class EarthEngine extends Layer {
...
-                options.projection
-            ) // eslint-disable-line
-        } else {
-            dictionary = this.eeImage.reduceRegion(
-                this._ee.Reducer.mean(),
-                point
-            ) // eslint-disable-line
-        }
+    // Returns value at at position
+    getValue = async latlng => {
```
[the remark this answers is not in the document]
Comment (turban) [reply src/layers/EarthEngine.js]: Good catch, the function returns a promise, so `await` is not needed. Removed.

**Pull request merged**